app/guided.py

```diff
@@ -0,0 +1,413 @@
+
+        self.half_life_penalty = half_life_penalty
+        self.half_life_gain = half_life_gain
+        self.goal_action_decay = goal_action_decay
+        self.goal_weight = goal_weight
+        self.fact_score_weight = fact_score_weight
+        self.last_action = None
+        self.goal_actions = set()
+
+        self.stopping_conditions = stopping_conditions
+        self.stopping_condition_met = False
+        self.state_machine = ['guided']
+        self.next_bucket = 'guided'
+
+    async def execute(self):
+        await self.planning_svc.execute_planner(self)
+    
+    async def guided(self):
+        await self.execute_subop(abilities=self.abilities, goals=self.goals)
+
+    async def execute_subop(self, abilities, agent=None, goals=[]):
+        """
+        :param abilities: List of ability IDs to plan with.
+        :param agent: Agent to focus planning on. If no agent is provided, planning will focus on all agents available in the operation.
+        :param goals: List of Goal objects to plan towards. Providing an empty list or an Exhaustion goal will cause the planner to infer goals from the ability's dependency graph.
+        """
+        attack_graph = await self._build_attack_graph(abilities, agent)
+
+        if len(goals) == 0 or goals[0].target == EXHAUSTION_KEY:
+            planner_goals = await self._create_terminal_goals(attack_graph)
+        else:
+            planner_goals = goals[:]
+
+        absolute_distance_table = await self._build_distance_table(attack_graph, goals=planner_goals)
+        effective_distance_table = copy.deepcopy(absolute_distance_table)
+
+        goal_links = await self._get_goal_links(absolute_distance_table, agent=agent)
+        while len(planner_goals) > 0 and len(goal_links) > 0:
+            tasked_links = await self._task_agents(goal_links, effective_distance_table, abilities, agent=agent)
+
+            await self.operation.wait_for_links_completion(tasked_links)
+            
+            current_goal_count = len(planner_goals)
+            planner_goals = await self._update_goals(current_goals=planner_goals)
+            if len(planner_goals) < current_goal_count:
+                absolute_distance_table, effective_distance_table = await self._handle_goal_achieved(attack_graph, goals=planner_goals)
+            else:
+                effective_distance_table = await self._handle_no_goal_achieved(absolute_distance_table,
+                                                                                effective_distance_table,
+                                                                                goal_links)
+            goal_links = await self._get_goal_links(absolute_distance_table, agent=agent)
+
+        self.planning_svc.log.debug('No more links available or all goals satisfied, planner complete.')
+        self.next_bucket = None
+
+    """ PRIVATE """
+
+    async def _build_attack_graph(self, abilities, agent):
```

Comment:
This is interesting. So this code will successfully reconcile common input/output facts and thus connect sequential abilities successfully?

Reply:
Yeah, finds connections in three ways. First is what it's referring to as input facts, which parses the traits out of the command templates, second is the output facts which is grabs from the parser configs, and third is from requirements definitions. Final graph should be contain nodes for all abilities in an adversary as well as nodes for each unique fact trait. There should be a directional connection from every ability to/from every fact trait that is used with it. Final graph would ideally allow you to determine whether there is any path between an input fact source and goals, and if there is what the critical path for that would be. 

Just as a note, I think longer term this is a good candidate for functionality to incorporate directly into the planning service in Caldera. It could be applied with the everything bagel adversary when using an initial fact source and goals to filter down abilities to run, so you wouldn't require an explicit adversary profile, but it still wouldn't do a complete search of action space.

Reply:
That is the plan.